File: dotnet/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/Models/Abstract/DocumentBase.cs
using System.Text.Json;

namespace PrototypeDesignPattern.Example1.Models.Abstract;

public abstract class DocumentBase : IDocumentPrototype<DocumentBase>
{
    public string Id { get; set; }
    public string Title { get; set; }
    public string Content { get; set; }
    public DateTime CreatedDate { get; set; }
    public DateTime LastModified { get; set; }
    public DocumentMetadata Metadata { get; set; }

    protected DocumentBase()
    {
        Id = Guid.NewGuid().ToString();
        CreatedDate = DateTime.Now;
        LastModified = DateTime.Now;
        Metadata = new DocumentMetadata();
    }

    // Shallow clone: reference-type properties share the same reference
    public virtual DocumentBase Clone()
    {
        var cloned = (DocumentBase)this.MemberwiseClone();
        cloned.Id = Guid.NewGuid().ToString();
        cloned.CreatedDate = DateTime.Now;
        cloned.LastModified = DateTime.Now;
        return cloned;
    }

    // Deep clone via JSON serialization: produces a fully independent copy
    public virtual DocumentBase DeepClone()
    {
        var json = JsonSerializer.Serialize(this, this.GetType());
        var cloned = (DocumentBase)JsonSerializer.Deserialize(json, this.GetType())!;
        cloned.Id = Guid.NewGuid().ToString();
        cloned.CreatedDate = DateTime.Now;
        cloned.LastModified = DateTime.Now;
        return cloned;
    }

    public virtual string GetDocumentInfo() =>
        $"ID: {Id}, Title: {Title}, Type: {GetType().Name}, Created: {CreatedDate:yyyy-MM-dd HH:mm}";

    public abstract bool ValidateDocument();
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/src/models/abstract/DocumentBase.ts
import { IDocumentPrototype } from './IDocumentPrototype';
import { DocumentMetadata } from '../DocumentMetadata';

export abstract class DocumentBase implements IDocumentPrototype<DocumentBase> {
  id: string;
  title: string = '';
  content: string = '';
  createdDate: Date;
  lastModified: Date;
  metadata: DocumentMetadata;

  protected constructor() {
    this.id = crypto.randomUUID();
    this.createdDate = new Date();
    this.lastModified = new Date();
    this.metadata = new DocumentMetadata();
  }

  // Shallow clone: reference-type properties share the same reference
  clone(): DocumentBase {
    const cloned = Object.assign(Object.create(Object.getPrototypeOf(this)), this) as DocumentBase;
    cloned.id = crypto.randomUUID();
    cloned.createdDate = new Date();
    cloned.lastModified = new Date();
    return cloned;
  }

  // Deep clone via JSON serialization: produces a fully independent copy
  deepClone(): DocumentBase {
    const cloned = JSON.parse(JSON.stringify(this)) as DocumentBase;
    Object.setPrototypeOf(cloned, Object.getPrototypeOf(this));
    cloned.id = crypto.randomUUID();
    cloned.createdDate = new Date();
    cloned.lastModified = new Date();
    return cloned;
  }

  getDocumentInfo(): string {
    return `ID: ${this.id.slice(0, 8)}..., Title: ${this.title}, Type: ${this.constructor.name}, Created: ${this.createdDate.toISOString().slice(0, 16)}`;
  }

  abstract validateDocument(): boolean;
}
